test/shared/src/main/scala/zio/test/ZIOSpec.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2021 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.test
+
+import org.portablescala.reflect.annotation.EnableReflectiveInstantiation
+import zio._
+import zio.test.environment.TestEnvironment
+import zio.test.render._
+
+@EnableReflectiveInstantiation
+abstract class ZIOSpec[R <: Has[_]: Tag](val layer: ZLayer[TestEnvironment, Any, R]) extends ZIOApp { self =>
+
+  def spec: ZSpec[R with TestEnvironment with Has[ZIOAppArgs], Any]
+
+  def aspects: Chunk[TestAspect[Nothing, R with TestEnvironment with Has[ZIOAppArgs], Nothing, Any]] =
```

Comment:
Combination of aspects is somewhat ill defined here. It doesn't really matter because a combined spec will run each individual spec with its own aspects, which is the right thing to do I think, but not clear what the interpretation of calling this on a combined spec should be. Perhaps it should be empty in the case of combined specs?

Reply:
If you think of this as a tree like structure and `aspects` as representing aspects at that level of the tree then empty actually makes sense here. When we combine two specs each spec may have its own aspects but there are no aspects that apply to the entire tree at that point. We could potentially add an operator allowing adding additional aspects to the new root.

Reply:
I agree that in the composition, this should be empty since the aspects have been pushed into the tests. Really, `aspects` is just a sort of "lazy application" of aspecs to those tests defined in `spec`, so this is the only sensible composition.